CMake/utility.cmake

```diff
@@ -386,6 +386,51 @@ function(activate_precompiled_headers _PRECOMPILED_HEADER _SOURCE_FILES)
   endif()
 endfunction()
 
+####################################################################################
+# Adds warnings compiler options to the target depending on the category
+# _TARGET Target name
+# _CATEGORY "library" or "unit_tests" or "benchmarks"
+####################################################################################
+macro( add_warning_levels _TARGET _CATEGORY)
+  if("${_CATEGORY}" STREQUAL "library" OR "${_CATEGORY}" STREQUAL "examples")
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+      set(CXX_FLAGS "-Werror")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+      set(CXX_FLAGS "-Werror")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+      set(CXX_FLAGS "/WX" "/W4")
+    else()
+      message(FATAL_ERROR "Unknown compiler to set warning levels: ${_CATEGORY} for target ${_TARGET}! ")
+    endif()
+  elseif("${_CATEGORY}" STREQUAL "unit_tests")
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+      set(CXX_FLAGS "-Werror" "-Wno-unused-private-field" "-Wno-delete-non-virtual-dtor")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+      set(CXX_FLAGS "-Werror" "-Wno-unused-private-field" "-Wno-delete-non-virtual-dtor")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+      set(CXX_FLAGS "/WX" "/W4")
+    else()
+      message(FATAL_ERROR "Unknown compiler to set warning levels: ${_CATEGORY} for target ${_TARGET}! ")
+    endif()
+  elseif("${_CATEGORY}" STREQUAL "benchmarks")
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+      set(CXX_FLAGS "-Werror" "-Wno-unused-private-field" "-Wno-delete-non-virtual-dtor")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+      set(CXX_FLAGS "-Werror" "-Wno-unused-private-field" "-Wno-delete-non-virtual-dtor")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+      set(CXX_FLAGS "/WX" "/W4")
+    else()
+      message(FATAL_ERROR "Unknown compiler to set warning levels: ${_CATEGORY} for target ${_TARGET}! ")
+    endif()
+  endif()
+
+  target_compile_options(${_TARGET} PRIVATE ${CXX_FLAGS} )
+  target_compile_options(${_TARGET} PRIVATE $<$<CONFIG:Debug>:${CXX_FLAGS_DEBUG}> )
```

Comment:
FYI: this call will be done automatically by CMake. So this call is redundant IMHO.
See my newest changes in my PR #107 , there I have fine grained your idea with the warnings and settings per target.

Reply:
Ah jeah, all my changes, except the individual warning levels are redundant! your changes in #107. Is   I think already ahead :)